### ISSUE TITLE
Enable `multi_commit` in `upload_folder` to work for branches/revisions

### ISSUE THREAD

**Issue by lewtun** [huggingface/huggingface_hub#1892]
**Is your feature request related to a problem? Please describe.**

I fine-tune a _lot_ of LLMs and to keep things organised I store the fine-tuned checkpoints as branches/revisions under a single model repo. For instance, we might have `mistral-7b-sft` as the model repo and then various branches (`v0.0`, `v0.1`, ...) storing the model weights associated with a given experiment.

One pain point for very large sharded models (i.e. those with >34b params) is that pushing the weights to the Hub is quite slow with the default settings of `upload_folder()`. 

Thanks to a tip from @Wauplin, I learned that one can speed this up by uploading the shards in parallel via the `multi_commit` feature ([docs](https://huggingface.co/docs/huggingface_hub/v0.19.3/guides/upload#upload-a-folder-by-chunks)).

However, this only works when uploading files to the `main` branch due to this check in the source code ([link](https://github.com/huggingface/huggingface_hub/blob/082c1ed895d1034cfeafd99c41218c99d8c65e8b/src/huggingface_hub/hf_api.py#L4435-L4437)).

**Describe the solution you'd like**

I would like `multi_commit` to also work for specific revisions, e.g. when calling:

```python
    upload_folder(
        repo_id="my-massive-model",
        folder_path="checkpoints",
        revision="v0.0",
        multi_commits=True,
        multi_commits_verbose=True,
    )
```

**Describe alternatives you've considered**

I've also experimented with `hf_transfer` to speed up the uploads and am aware of `plan_multiple_commits()` as a possible workaround. What I'm unsure of is whether `plan_multiple_commits()` would hit the same limitation as needed to open a PR on a specific revision, before merging.

**Additional context**

**Comment by lewtun** [comment]
Ah, thanks a lot for the clarification about `multi_commit=True`! I naively assumed it pushed the shards asynchronously to parallelise the upload - which would be kind of nifty, but probably too complex :)

I haven't experienced too many situations where a retry was needed, so I don't think this warrants a change to the codebase just for 1 user :)

Thanks again and feel free to close the issue.

**Comment by Wauplin** [comment]
> I naively assumed it pushed the shards asynchronously to parallelise the upload - which would be kind of nifty, but probably too complex :)

Shards are uploaded in parallel by default when using `upload_folder`! `hf_transfer` makes it even faster by parallelizing uploads of file chunks. 

I'll close this issue as "won't fix" but if anyone land here and would like to see this feature implemented, please comment. If there's more demand, we'll find a way to do it :)
